python/mxnet/ndarray/ndarray.py

```diff
@@ -507,71 +537,269 @@ def __getitem__(self, key):
-                shape = None
-                for idx_i in key:
-                    if not isinstance(idx_i, NDArray):
-                        assert isinstance(idx_i, (NDArray, np.ndarray, list, tuple)), \
-                            "Combining basic and advanced indexing is not supported " \
-                            "yet. Indices must be all NDArray or all slice, not a " \
-                            "mix of both."
-                        idx_i = array(idx_i, ctx=self.context, dtype=dtype)
-                    else:
-                        dtype = idx_i.dtype
-                    if shape is None:
-                        shape = idx_i.shape
+            else:
+                return self
+
+        # Start dealing with more complicated indexing cases than only slicing at axis=0.
+        # if key is any type of NDArray, list, or np.ndarray make a tuple from it
+        # Note: Only support integer list as index if key is a list.
+        if isinstance(key, (NDArray, np.ndarray)):
+            key = (key,)
+        elif isinstance(key, list):
+            # TODO(junwu): Add support for nested lists besides integer list
+            for i in key:
+                if not isinstance(i, integer_types):
+                    raise TypeError('Indexing NDArray only supports a list of integers as index'
+                                    ' when key is of list type, received element=%s of type=%s'
+                                    % (str(i), str(type(i))))
+            key = (key,)
+
+        if isinstance(key, tuple):
+            indexing_dispatch_code = _get_indexing_dispatch_code(key)
+            if indexing_dispatch_code == _BASIC_INDEXING_STEP_EQUAL_ONE:
+                return self._basic_indexing_step_equal_one(key)
+            elif indexing_dispatch_code == _BASIC_INDEXING_STEP_NOT_ONE:
+                return self._basic_indexing_step_not_one(key)
+            elif indexing_dispatch_code == _ADVANCED_INDEXING:
+                return self._advanced_indexing(key)
+            else:
+                raise ValueError('Indexing NDArray with index=%s is not supported' % str(key))
+        else:
+            raise ValueError('NDArray does not support slicing with key %s of type %s'
+                             % (str(key), str(type(key))))
+
+    def _advanced_indexing(self, key):
+        """Get item when key is a tuple of any objects of the following types:
+        NDArray, np.ndarray, list, tuple, slice, and integer. Note that if
+        all the objects in the tuple are of type slice with step=1, it should
+        invoke _basic_indexing_step_equal_one for better efficiency."""
+        def _is_advanced_index(index):
+            """The definition of advanced index here includes integers as well, while
+            integers are considered as basic index type when the key contains only
+            slices and integers."""
+            return not isinstance(index, py_slice)
+
+        if not isinstance(key, tuple):
```

Comment:
Replace with simple assert since this is never reached.

Reply:
Done.